Support ES6 module syntax in Acorn and Closure

kleisauke opened the pull request:
Acorn and Closure are now made aware of ES6 module syntax when
linking with `-sEXPORT_ES6`. The previous mangling of top-level
await, `import.meta` and dynamic `import()` is tightened to only
mangle top-level await when Closure is enabled, since this is the
only construct not supported there.

kleisauke committed:
Cherry-pick 234940f032e4353a4390af02bbc59a6a9f87afcf

kleisauke committed:
Fix test failure in `core2.test_modularize_closure_pre`

kleisauke committed:
Fix test failure in `core2.test_pthread_proxying_modularize`

sbc100 commented on test/code_size/hello_webgl_wasm.json:
```diff
@@ -1,10 +1,10 @@
   "a.html.gz": 328,
-  "a.js": 3920,
-  "a.js.gz": 2089,
+  "a.js": 5629,
```
This is pretty big regression sadly.

kleisauke replied:
Here's the consolidated and prettified diff between main and this PR for `other.test_minimal_runtime_code_size_hello_webgl_wasm`:

<details>
  <summary>Details</summary>

```diff
--- a/a.js
+++ b/a.js
@@ -1,4 +1,4 @@
-var Module = async function (B = {}) {
+async function Module(B = {}) {
   var l,
     p,
     r,
@@ -7,6 +7,88 @@ var Module = async function (B = {}) {
     y,
     D,
     E,
+    h = {
+      xa: 0,
+      Ha: 0,
+      Ka: 0,
+      Qa: 0,
+      hb: 0,
+      eb: 0,
+      Ea: 0,
+      Ca: 0,
+      Sa: 0,
+      Ba: 0,
+      Ga: 0,
+      Ta: 0,
+      gb: 0,
+      cb: 0,
+      Oa(a, b, c) {
+        r.set(r.subarray(b, b + c), a);
+      },
+      Ya() {
+        for (; h.L.length; ) h.V(h.L.length - 1);
+        h.N = [];
+      },
+      X: 0,
+      N: [],
+      za(a, b, c) {
+        function d(k, p) {
+          if (k.length != p.length) return !1;
+          for (var q in k) if (k[q] != p[q]) return !1;
+          return !0;
+        }
+
+        for (var f of h.N) if (f.Y == a && d(f.$, c)) return;
+        h.N.push({Y: a, da: b, $: c});
+        h.N.sort((k, p) => k.da < p.da);
+      },
+      Za(a) {
+        h.N = h.N.filter((b) => b.Y != a);
+      },
+      ja() {
+        return navigator.userActivation ? navigator.userActivation.isActive : h.X && h.la.wa;
+      },
+      ea() {
+        if (h.ja()) {
+          var a = h.N;
+          h.N = [];
+          for (var b of a) b.Y(...b.$);
+        }
+      },
+      L: [],
+      sa: (a, b) => {
+        for (var c = 0; c < h.L.length; ++c) h.L[c].target != a || (b && b != h.L[c].T) || h.V(c--);
+      },
+      V(a) {
+        var b = h.L[a];
+        b.target.removeEventListener(b.T, b.ba, b.va);
+        h.L.splice(a, 1);
+      },
+      Xa(a) {
+        if (!a.target) return -4;
+        if (a.ya)
+          (a.ba = function (c) {
+            ++h.X;
+            h.la = a;
+            h.ea();
+            a.Ja(c);
+            h.ea();
+            --h.X;
+          }),
+            a.target.addEventListener(a.T, a.ba, a.va),
+            h.L.push(a);
+        else
+          for (var b = 0; b < h.L.length; ++b)
+            h.L[b].target == a.target && h.L[b].T == a.T && h.V(b--);
+        return 0;
+      },
+      Ia(a) {
+        return a ? (a == window ? '#window' : a == screen ? '#screen' : a?.nodeName || '') : '';
+      },
+      fullscreenEnabled() {
+        return document.fullscreenEnabled || document.webkitFullscreenEnabled;
+      },
+    },
     M = new TextDecoder(),
     w = (a, b) => {
       if (!a) return '';
@@ -283,4 +365,8 @@ var Module = async function (B = {}) {
     L();
   });
   return {};
-};
+}
+
+'object' === typeof exports && 'object' === typeof module
+  ? ((module.exports = Module), (module.exports.default = Module))
+  : 'function' === typeof define && define.amd && define([], () => Module);
```
</details>


One factor is that the CJS/AMD module exports are now present under `MINIMAL_RUNTIME` (see PR #23282). The other factor is likely(?) that JSDCE is unable to remove certain parts of `libhtml5.js`.

sbc100 replied:
I think we should try to land f53a92d as its own patch maybe?

kleisauke replied:
PR #23850 resolves this size regression instead. 

It seems that `other.test_minimal_runtime_code_size_hello_webgl_wasm` depends on:
- `emscripten_set_element_css_size()`
- `emscripten_set_canvas_element_size()`
- `emscripten_webgl_create_context()`

All of these had an unused dependency on `$JSEvents`, which, for some reason, JSDCE fails to remove in this changeset.

sbc100 replied:
Are you sure its JSDCE and not closure compile that is failing to remove these things?

That was my experience too in #23261.. I figured the optimizers were just less good at removing non-top-level code?

kleisauke replied:
Closure might also be failing to remove this unused dependency - I haven't checked.

We could try replacing Closure with https://github.com/swc-project/swc, as it shows promising results in https://github.com/privatenumber/minification-benchmarks. However, I'm unsure how much effort that would require.

sbc100 replied:
I would *love* to replace closure with pretty much anything else :)

However, so far nothing as come along that can't do as good of a job at minifying.   Maybe its worth looking into swc!

kleisauke committed:
Remove redundant code

This file is not included when building with `-sMODULARIZE=instance`.

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (6) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2504 => 2427 [-77 bytes / -3.08%]
codesize/test_minimal_runtime_code_size_hello_webgl2_wasm.json: 13199 => 13656 [+457 bytes / +3.46%]
codesize/test_minimal_runtime_code_size_hello_webgl2_wasm2js.json: 18537 => 19154 [+617 bytes / +3.33%]
codesize/test_minimal_runtime_code_size_hello_webgl2_wasm_singlefile.json: 15132 => 15604 [+472 bytes / +3.12%]
codesize/test_minimal_runtime_code_size_hello_webgl_wasm.json: 12737 => 13163 [+426 bytes / +3.34%]
codesize/test_minimal_runtime_code_size_hello_webgl_wasm2js.json: 18063 => 18652 [+589 bytes / +3.26%]

Average change: +2.24% (-3.08% - +3.46%)
```

kleisauke committed:
Fix test failure in `browser.test_cross_origin_es6`

kleisauke committed:
Ensure 2-space indent

kleisauke committed:
Fix test failure in `browser_2gb.test_source_phase_imports`

kleisauke committed:
Avoid dependence on PR #23261

kleisauke committed:
Closure doesn't support top-level await

kleisauke committed:
Allow Babel to parse top-level await

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2504 => 2474 [-30 bytes / -1.20%]

Average change: -1.20% (-1.20% - -1.20%)
```

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Revise comments

kleisauke committed:
Allow Babel to parse `import.meta`

kleisauke committed:
Ensure `import()` expressions are not Babeled

kleisauke committed:
Ensure Closure doesn't minify `require()`

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2494 => 2525 [+31 bytes / +1.24%]

Average change: +1.24% (+1.24% - +1.24%)
```

kleisauke committed:
`globalThis` -> `global`

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2525 => 2521 [-4 bytes / -0.16%]

Average change: -0.16% (-0.16% - -0.16%)
```

kleisauke committed:
Add comment regarding `export{};`

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2523 => 2520 [-3 bytes / -0.12%]

Average change: -0.12% (-0.12% - -0.12%)
```

kleisauke committed:
`modules.js` was renamed to `modules.mjs`

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Revert "Ensure Closure doesn't minify `require()`"

This reverts commit ef36474b9347f074f01dbdbe6be96827413be2ef.

kleisauke committed:
Restore the mangling logic to its original location

kleisauke committed:
Prefer use of `replaceAll`

kleisauke committed:
Add comment

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2520 => 2493 [-27 bytes / -1.07%]

Average change: -1.07% (-1.07% - -1.07%)
```

kleisauke committed:
`parseTools.js` was renamed to `parseTools.mjs`

sbc100 commented on tools/emscripten.py:
```diff
@@ -427,6 +427,11 @@ def emscript(in_wasm, out_wasm, outfile_js, js_syms, finalize=True, base_metadat
+  # Unmangle previously mangled `import.meta` references in lib*.js.
+  # See also: `LibraryManager.load` in modules.mjs.
+  if settings.EXPORT_ES6:
+    glue = glue.replace('EMSCRIPTEN$IMPORT$META', 'import.meta')
```
Maybe we could run this transform over all of the `libraryItems` before we call writeOutput?

kleisauke committed:
Move unmangling to JS compiler

kleisauke committed:
Update comment

kleisauke commented on tools/emscripten.py:
```diff
@@ -427,6 +427,13 @@ def emscript(in_wasm, out_wasm, outfile_js, js_syms, finalize=True, base_metadat
+  # Unmangle previously mangled `import.meta` references in lib*.js.
+  # See also: `preprocess` in parseTools.mjs.
+  # Note: this needs to be done here because this is the only stage
+  # where the complete file contents are available.
```
`import.meta` is supported by Acorn, Babel and Terser when `allowImportExportEverywhere` is enabled and by Closure when using `--chunk_output_type ES_MODULES`. So, this (un-)mangling is only applied for `vm.runInContext`, where `import.meta` is not supported.

`fix_js_mangling` is now only used under Closure in `MODULARIZE` mode, which is a nice win. Once https://github.com/google/closure-compiler/issues/3835 is fixed, we should be able to remove this Closure-specific (un-)mangling for top-level await entirely.

kleisauke committed:
Update comment (2)

kleisauke commented:
Note to self: PR #26677 would make the Babel changes unnecessary.

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2521 => 2491 [-30 bytes / -1.19%]

Average change: -1.19% (-1.19% - -1.19%)
```

kleisauke committed:
Merge branch 'main' into es6-acorn-closure

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_minimal_esm.json: 2508 => 2414 [-94 bytes / -3.75%]

Average change: -3.75% (-3.75% - -3.75%)
```

kleisauke committed:
Ensure Closure doesn't minify `Module` in `--chunk_output_type=ES_MODULES` mode

Fixes a test failure in `other.test_modularize_incoming_closure`.

sbc100 commented on src/jsifier.mjs:
```diff
@@ -397,6 +397,12 @@ export async function runJSify(outputFile, symbolsOnly) {
 
   async function writeOutput(str) {
+    // Unmangle previously mangled `import.meta` references.
+    // See also: `preprocess` in parseTools.mjs.
```
The mangling was moved to `processMacros` and its now done by a function called `mangleUnsupportedSyntax` as of #27059.  Perhaps reference that new function name?

kleisauke replied:
Goot catch! I forgot to do this within the merge commit, fixed with 10e638306bfb5b19ad0050f0328de1ac854e2af8.

sbc100 commented on src/jsifier.mjs:
```diff
@@ -397,6 +397,12 @@ export async function runJSify(outputFile, symbolsOnly) {
+    // See also: `preprocess` in parseTools.mjs.
+    if (EXPORT_ES6) {
+      str = str.replaceAll('EMSCRIPTEN$IMPORT$META', 'import.meta');
+    }
```
Perhaps mention why we are doing this?  

Something like "Unmangle import.meta which was only mangled to allow the code to flow through vm.runInContext".

kleisauke replied:
This is already mentioned in the comment that is referenced.
https://github.com/emscripten-core/emscripten/blob/10e638306bfb5b19ad0050f0328de1ac854e2af8/src/parseTools.mjs#L38-L43

sbc100 commented on tools/link.py:
```diff
@@ -2073,19 +2073,28 @@ def phase_source_transforms(options):
+# Unmangle previously mangled `await import` and `await` references in
 # both main code and libraries.
-# See also: `preprocess` in parseTools.js.
+# See also: `preprocess` in parseTools.mjs.
```
`preprocess` -> `mangleUnsupportedSyntax`

kleisauke replied:
10e638306bfb5b19ad0050f0328de1ac854e2af8

sbc100 commented on tools/link.py:
```diff
@@ -2073,19 +2073,28 @@ def phase_source_transforms(options):
+    # https://github.com/google/closure-compiler/blob/v20260401/src/com/google/javascript/jscomp/ConvertChunksToESModules.java#L111-L113
+    src = src \
+      .replace('EMSCRIPTEN$AWAIT$IMPORT', 'await import') \
+      .replace('export{};\n', '')
```
Should we at `^` to the start of the `export{}` line to be more precise?

kleisauke replied:
That will require use of `re.sub`, I thought this was simpler.

sbc100 replied:
Fair enough

sbc100 commented:
I'm excited to see this land!

kleisauke committed:
Update comment references
